Better IME support

shibukawa opened the pull request:
Add better IME support (#41) to implement on-the-spot IME conversion. My code is based on #643.

I added the following functions:
- glfwSetPreeditCallback() : callback function to notify preedit text
- glfwSetPreeditCursorPos() : teach text cursor position for IME windows
- glfwGetPreeditCursorPos() : get text cursor position for IME windows

They are not big changes. It keeps backward compatibility about API level.
To use IME feature, UI programmers should implement visualization code for preedit texts. 

I implemented platform specific code for Windows/X11/Mac OS X (not Wayland and Mir).

shibukawa commented:
I start implementing my text box, I need some features in addition to the above functions:

* Set/Get current IME status (integer only text box and so on)
* IME status change callback (commit text when turn off IME)
* Reset preedit text (reset text when widget focus would changed)

I will add these functions to my branch.

elmindreda commented:
My initial impression is that this looks great! I will review it more closely once it's time for the IME part of 3.2, which should hopefully give you time to land the additional features you mention.

shibukawa commented:
I implemented all method I mentioned and I added one test application (ime.c, almost as same as event.c, but IME functions are assigned to mouse buttons).

Current status is:

<img width="317" alt="glfw-screenshot" src="https://cloud.githubusercontent.com/assets/564612/11846456/1bfa2690-a45c-11e5-827d-52a374d1cc60.png">

* Windows works almost perfectly except Google Japanese Input doesn't follow cursor position (MS IME works).
* Mac OS X's NSTextInputContext doesn't work on my computer (10.9). So resetting preedit text feature doesn't work.
* I tested on Ubuntu 15.10, but XIM (X Input Method) implementation is buggy. Test application can receive preedit text and commited text, but my real application can't. I am still debugging.

XIM is very old standard. I think adding Fcitx option (more modern protocol and most popular one now) is better solution (use Fcitx if it is installed, otherwise use XIM). And Fcitx supports Wayland too (I couldn't find Mir information). How about it?

elmindreda commented:
There is quite a range of programming styles in this patch. Did you write this yourself or are parts of it copied from other sources? If so I need to know what licenses they come under.

elmindreda commented:
The `glfwGetIMEStatus` and `glfwSetIMEStatus` functions should be folded into `glfwGetInputMode` and `glfwSetInputMode`.

elmindreda commented:
The `imm32` library should be dynamically loaded the same way as other non-default libraries.

shibukawa commented:
Maybe 50% is my code and 50% is other's. I referred the following Japanese web sites:

Windows: http://www.kumei.ne.jp/c_lang/sdk3/sdk_278.htm
X11: http://homepage3.nifty.com/rio_i/lab/xlib/012xim.htm

I will ask them about sample code's license.

I couldn't find memo about Mac. Of course, I checked MSDN, Apple's document, X11R6 references.

shibukawa commented:
I will merge glfw(Get|Set)IMEStatus into glfw(Get|Set)InputMode.

imm32 is provided as static library by Microsoft and it should be bundled with Windows like gdi32.dll. I thinks if(WIN32) is enough to detect imm32.

shibukawa commented:
1. I rebased to fix conflict between current master.
2. I removed glfw(Get|Set)IMEStatus and added new options to glfw(Get|Set)InputMode
3. Ryo Ichinose (X11 guy) agreed using XIM related code as libpng/zlib license.

Now I am waiting reply from Yasutaka Kumei.

shibukawa commented:
Yasutaka Kumei agreed using his code with glfw.

elmindreda commented:
Awesome! Thank you!

elmindreda commented:
About the `imm32` thing, it's not for compatibility with older systems. As you rightly point out, the import library is available everywhere. Rather it's as a convenience for users, just like the current dynamic loading of `winmm`.

hesiod commented:
@elmindreda It appears that you wanted to close this PR with 779b562, but Github failed to realize that, so you probably have to close this PR manually

shurcooL commented:
> It appears that you wanted to close this PR with 779b562, but Github failed to realize that, so you probably have to close this PR manually

That commit is not yet merged into `master` branch. When it is, due to its commit message, this PR will be closed.

elmindreda commented:
What @shurcooL said. Expect several more revisions to pop up here before it's ready to merge, as it's a big and multi-platform addition.

hajimehoshi commented:
What is going on this?

molikto commented:
Can anyone comment what's needed to get this merged?

shibukawa committed:
fixup ef70fee8 Add preedit text callback API and implementation, IME status API for X11, macOS, Windows

durswd commented:
I merged the latest version and fix a bug based on shibukawa-san branch.
https://github.com/altseed/glfw/tree/feature/add-ime-support

When is this PR merged?
@elmindreda 
What do I do to merge this PR (Add documentation, fix bugs, etc)?

Windmill-City commented:
Imm32 is problematic on Window10 as Microsoft has dropped support of it.
It's better to use Text Service Framework instead.
And there should be a way to get candidate list, as input method can't show its candidate list window when the game is in fullscreen mode

daipom commented:
I would like to continue this fix with #2117.
I have already contacted Mr. @shibukawa .

ashie commented:
It appears that #2130 now fully incorporates the work from this PR, including attribution to the original contributors.

As far as I can tell, there is no remaining work in this PR that is not already represented in #2130. In that sense, #2130 seems to have effectively superseded this PR.

It may therefore make sense to close this PR and continue any future discussion and review efforts on #2130 instead.

Thank you for the original contribution. The work done here laid the foundation for the current effort, and it is good to see it being carried forward.
